Mitmproxy-based automated failure testing

lithp opened the pull request:
Some slides explaining how this works: https://docs.google.com/presentation/d/1zxF32GvcFJp0s6UClL39hla9WqhbGvYgo30e2WYnADQ/edit?usp=sharing

A modernized #2044. This one works at the network level instead of the process level. A drawback is: we don't have a good way of exposing timing problems. An example might be, maybe if a packet comes in before the process is ready for it the process crashes. That class of bugs seems rare to me, so I think this approach is safe to move forward with.

It's also much easier! This doesn't require sudo, just an extra daemon which packets flow through. It also has the advantage of not caring at all how the process runs, so the same tests should work modulo any refactoring which we choose to do.

Some work left to be done:

- [x] There's a race condition which causes mitmproxy to sometimes return `'NoneType' object has no attribute 'tell'`. What is going wrong?
- [ ] Should we really be using a fifo? Maybe a socket makes more sense? A hostname:port is much more portable (though we would still need some way of blocking to make sure changes have been made, and would have to rely on a tool such as `netcat` in order to keep using `COPY`)
- [ ] Refactor fluent.py into something more maintainable, the double-queue solution probably isn't optimal.
- [ ] The `Handler` and `Mixin` hierarchy is cute but there's definitely a better way. Separating out the Building and the Handling seems like it'd improve things.
- [ ] Convert all of our manual tests into automated tests under this framework!
- [ ] Either `pg_regress_multi.pl` or `fluent.py` should redirect the proxy output to a log file
- [x] How hard would it be to extend this to cancellation testing?
- [x] Instead of bumping the Citus version, add `citus.mitmproxy()` to the test functions?
- [ ] Currently it's only possible to fail one of the workers, should we add a second mitmproxy so we can play with that worker as well?
- [ ] `DROP TABLE` crashes if you drop the connection after `DROP TABLE IF EXISTS` is sent. Open a ticket!
- [ ] What do we do about test output which includes PIDs?
- [ ] Profile this and figure out how to improve performance? 6 tests take 10 seconds to run.
- [x] `recorder.dump()` returns an empty line, fix that!
- [ ] Some of these tests could be made faster. Instead of recreating tables with different settings of shard_replication_factor, we can start by using the higher replication factor, then manually set some of the shards inactive.
- [x] Make failure_drop_table consistent, prepared transactions have random ids, introduce something like `enable_unique_job_ids` -> `enable_unique_transaction_ids`
- [x] Make the `failure_task_tracker_executor` test reproducible (don't hard-code filepaths)

lithp commented:
`COPY` with a `mkfifo` works great, but it relies on https://github.com/citusdata/citus/issues/2127 to somehow be fixed.

lithp commented:
It's finally running (albeit in a rather hacky way) and passing on Travis!

lithp commented:
Rebased onto master

lithp commented:
Clearing out some outdated TODOs:

- [x] rebase onto master
- [x] Jason brought up a good point: We could remove the dependency on `plperlu` by using `COPY` to write to our fifo.
- [x] The fifo location is hard-coded, use some kind of path which always works. Maybe make it a relative one?
- [x] Does the travis postgres have `plperlu`? Is that really something we want to be adding a dependency on?
- [x] Break out the tests into a new make target: `make check-failure`
- [x] Somehow install mitmproxy on travis
- [x] Have `pg_regress_multi.pl` start the mitmproxy daemon.
- [ ] Maybe we can work around #2127 by using `EXECUTE` so it plans the query afresh every time?

lithp commented:
#### Cancellation testing

The proxy sits at the network layer instead of acting directly on the binary. This means it's more stable under refactoring, but also means we have much less control over what the binary is doing when we take some action. If the action is sending a signal (to cancel the binary), it's likely that precise timing will be more important. We'll want to send a signal right at some critical section. So, maybe cancellation testing is better handled a different way. 

For instance, we could embed macro calls like

    SIGNAL_INJECTION_POINT("after_begin")

which raise a signal if the test script has previously called 

    citus.interrupt_at('after_begin')

### With mitmproxy

If mitmproxy is told the name of the process, it could send a signal when asked to. We could embed a script like `flow.contains(b"SELECT").interrupt(19234)`, which you would call with the result of `pg_backend_pid()`.

lithp commented:
[Here's the failure testing portion of our 7.4 release testing plan.](https://docs.google.com/document/d/1sseiMEPkA3dDqgNrFypcY7f2CmkGl6fgMDC1yYmZCzU/edit#bookmark=id.a9e4djl23cmx)

lithp commented:
Here's the traceback when we the race condition happens:

```
Traceback (most recent call last):
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/bin/mitmdump", line 11, in <module>
    sys.exit(mitmdump())
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/tools/main.py", line 155, in mitmdump
    m = run(dump.DumpMaster, cmdline.mitmdump, args, extra)
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/tools/main.py", line 122, in run
    master.run()
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/master.py", line 93, in run
    self.tick(0.1)
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/master.py", line 109, in tick
    self.addons.handle_lifecycle(mtype, obj)
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/addonmanager.py", line 238, in handle_lifecycle
    self.trigger(name, message)
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/addonmanager.py", line 282, in trigger
    with safecall():
  File "/usr/lib/python3.5/contextlib.py", line 59, in __enter__
    return next(self.gen)
  File "/home/brian/.local/share/virtualenvs/regress-MQrw9tUE/lib/python3.5/site-packages/mitmproxy/addonmanager.py", line 60, in safecall
    tell = ctx.master.tell
AttributeError: 'NoneType' object has no attribute 'tell'
```

lithp commented:
All the tests from our test document:

- [ ] Fail some nodes while running a query, but all shards have at least one active placements
  - [ ] Do this with real time executor 
    - [ ] aggregates, simple joins, table fetching queries, functions
  - [ ] Do this with task tracker executor queries (Look at the task tracker design document for identifying/testing various failure scenarios)
    - [ ] aggregates, simple joins, re-partition joins 
  - [ ] Do this with router executor queries 
    - [ ] Aggregates, window functions, simple-key value lookups
- [ ] Fail some nodes while running a query, one shard doesn’t have any active placements
- [ ] COPY on distributed tables with (1PC)
  - [ ] Test with hash, append and range distributed tables 
  - [ ] Stop a worker before COPY starts;  
    - [ ] If there is one replica; COPY should error out, placement should stay healthy. 
    - [ ] If there are more than one replica; COPY should not error out, placements should be marked as INACTIVE. (This test is only applicable for hash distributed tables, otherwise the query errors out) 
  - [ ] Stop a worker during COPY;  
    - [ ] If there is one replica; COPY should error out, placement should stay healthy.  
    - [ ] If there are more than one replica; COPY should error out, placement should stay healthy. 
  - [ ] Stop a worker during COMMIT; 
    - [ ] If there is one replica; COPY should give warning, some of the data already committed, placements should stay healthy. 
    - [ ] If there are more than one replica; COPY should not error out, placement should be marked as INACTIVE.
    - [ ] If data is corrupted (i.e., 1 row is erroneous), COPY errors-out and no data is ingested to the workers 
- [ ] COPY on hash partitioned tables with default values (2PC):
  - [ ] Stop a worker before COPY starts;  
    - [ ] If there is one replica; COPY should error out, placement should stay healthy.
    - [ ] If there are more than one replica; COPY should not error out, placements should be marked as INACTIVE. 
  - [ ] Stop a worker during COPY;   
    - [ ] If there is one replica; COPY should error out, placement should stay healthy.
    - [ ] If there are more than one replica; COPY should error out, placement should stay healthy.
  - [ ] Stop a worker during COMMIT; 
    - [ ] If there is one replica; COPY should give warning, 
      - [ ] Wait show citus.recover_2pc_interval and see that prepared transactions are rollbacked on the workers via: SELECT run_command_on_workers('SELECT count(*) FROM pg_prepared_xacts');
      - [ ] You can mostly do that with gdb, stopping at StartRemoteTransactionCommit() on the calls after the first one
    - [ ] If there are more than one replica; COPY should not error out, placement should be marked as INACTIVE
      - [ ] I think this should be the expected behaviour
      - [ ] Citus errors out whole transaction and doesn’t mark any placements INVALID. (But, the problem is that we’re not consistent with the other cases such as when the worker is down at the beginning of the transaction )
      - [ ] Opened an issue for discussion #1933
- [ ] Copy on reference tables
  - [ ] Always use 2PC, never mark any placement as INACTIVE 
    - [ ] Example: If a table has unique constraint and we COPY the same row twice, COPY errors-out and no data is ingested to the workers and no placement is marked as inactive. 
- [ ] COPY on append partitioned failures
  - [ ] Shard creation fails: Create table with composite type and don’t create composite type on one node. Verify COPY handles failure as expected. 
  - [ ] Min/Max fetching fails: In the above don’t create min/max function for composite type on workers. Verify COPY handles failure as expected . 
  - [ ] Test COPY in transaction & ROLLBACK 
- [ ] Fail nodes during master_create_empty_shard
  - [ ] Fail one node and verify that it is able to create a shard. Verify metadata. 
  - [ ] Fail several nodes such that active nodes < shard_replication_factor and verify that no metadata is inserted. 
  - [ ] Test master_create_empty_shard in transaction & ROLLBACK
- [ ] Fail nodes during master_append_table_to_shard
  - [ ] Fail all nodes having the shard and verify that we error out and no metadata is added 
  - [ ] Fail some nodes having the shard and verify it is marked as invalid and metadata is updated accordingly 
  - [ ] Ctrl-C during master_append_table_to_shard 
  - [ ] Test master_append_table_to_shard in transaction & ROLLBACK 
- [ ] Fail nodes during master_apply_delete_command ( )
  - [ ] Verify that the transaction fails, no metadata changes are happened and none of the shard placements are dropped at all
  - [ ] Verify that pg_dist_shard_placement row is not deleted.
  - [ ] Verify that shards on nodes which are up are not deleted and their metadata is not removed.
  - [ ] Test  master_apply_delete_command & ROLLBACK
- [ ] Fail some nodes while running a subselect
  - [ ] behaviour should be same as executor behaviour  
    - [ ] Task-tracker  
    - [ ] Real-time 
    - [ ] Pull-push 
      - [ ] https://github.com/citusdata/citus/issues/2048
  - [ ] run a single table subquery, single node failure is tolerated, 2 node failure caused query to fail 
- [ ] Fail some nodes while inserting/updating/deleting rows
  - [ ] Test  BEGIN, inserting/updating/deleting rows, ROLLBACK 
  - [ ] Also test multiple shard inserts
- [ ] Fail on create_distributed_table
  - [ ] The whole transaction should be rollbacked in any case: With replication factor 1 or greater and with 2PC/1PC enabled, inside a transaction or not 
  - [ ] Test BEGIN; create_worker_shards(); ROLLBACK 
- [ ] Fail a node during create_distributed_table(..., colocate_with => '...')
  - [ ] Test with all colocate_with options 
- [ ] Fail nodes during create_reference_table()
- [ ] Fail nodes during DDL command propagation
  - [ ] Verify that if node is failed before executing DDL command, then command can’t be executed. 
  - [ ] Verify that if node fails during execution or if you get a postgres error on a particular shard and not the others, we error out and ROLLBACK. 
- [ ] Fail nodes during truncate command
- [ ] Fail nodes during insert into/select command
- [ ] Fail nodes during insert into/select command (when data is pulled to master)
- [ ] Fail nodes during add/remove/disable node with and without reference tables
- [ ] Fail nodes during modification to reference tables. Whole transaction should fail, no data should be inserted and no metadata update should happen.
- [ ] Fail a node during MX metadata sync
  - [ ] Test during create_distributed_table, ddl changes, mark_tables_colocated, DROP TABLE 
- [ ] Fail a node during creating distributed table from local NON-EMPTY table
  - [ ] Check metadata, check if workers have corrupted data etc. 
- [ ] Fail a node during creating distributed table from partitioned table
  - [ ] Check metadata, check if workers have corrupted data etc.
- [ ] Fail a node when creating a partition of a distributed-partitioned table
  - [ ] Check metadata, check if workers have corrupted data etc.
- [ ] Fail a MX node during DROP SEQUENCE
- [ ] Fail a node while running table size functions
  - [ ] This is probably okay. Size functions do not run when replication factor is greater than 1. They also fail when a single node is down.
- [ ] Fail a node during ALTER EXTENSION citus UPDATE
- [ ] Fail a node during master_add_node, master_activate_node, master_disable_node, master_add_inactive_node
  - [ ] master_add_inactive_node should not error out 
  - [ ] check metadata of reference tables, corruption of reference table data, replication count in pg_dist_colocation 
- [ ] Fail a node during CREATE INDEX … CONCURRENTLY
  - [ ] Issued created index concurrently
  - [ ] Killed a worker process
- [ ] Fail a node during ALTER TABLE ADD/DROP CONSTRAINT
- [ ] Fail a node during ALTER TABLE RENAME COLUMN
- [ ] Fail a node when creating a savepoint (or rollbacking to a savepoint)
  - [ ] When a node fails, the whole transaction fails anyway. But, keep the test anyway.
- [ ] Fail node when DISTINCT/COUNT DISTINCT running
  - [ ] Fail a node when processing is on the worker
  - [ ] Fail the coordinator when the processing is on the coordinator 
- [ ] Fail/kill maintenance daemon
  - [ ] When the daemon is sleeping 
  - [ ] When the daemon is running call home 
  - [ ] When the daemon is running deadlock detection 
  - [ ] When the daemon is running recovery 
- [ ] Fail node while recover_prepared_statements() in progress
  - [ ] Fail the coordinator 
  - [ ] Fail the workers 
- [ ] Fail node while running multi-shard UPDATE/DELETEs
  - [ ] Test with both 1PC and 2PC
    - [ ] Fail coordinator 
    - [ ] Fail worker

lithp commented:
Things to play with: `citus.shard_count`, `citus.shard_replication_factor`

lithp commented:
Useful: `get_shard_id_for_distribution_column(table-name, value)`

lithp commented:
How do you tell which executor is being used, to know which one a specific join uses?

lithp commented:
There are additional tests here: https://docs.google.com/document/d/10mEwsYftun6ONoRCmtwlNcKvgdHEuGzqTuBnAZJuUag/edit

lithp commented:
Can you add any test cases for this: https://github.com/citusdata/citus/pull/2031

lithp commented on src/test/regress/expected/failure_1pc_copy.out:
```diff
@@ -0,0 +1,107 @@
+WARNING:  connection not open
+CONTEXT:  while executing command on localhost:57640
+COPY copy_test, line 1: "0, 0"
+ERROR:  failure on connection marked as essential: localhost:57640
```
I think this is a bug. If the other worker is accessible we shouldn't error out, we should just mark the placement inactive

lithp commented on src/test/regress/expected/failure_1pc_copy.out:
```diff
@@ -0,0 +1,107 @@
+	This probably means the server terminated abnormally
+	before or while processing the request.
+CONTEXT:  while executing command on localhost:57640
+COPY copy_test, line 1: "0, 0"
```
Same here, the other worker is doing fine, this should mark the placement inactive and continue.

lithp commented on src/test/regress/expected/failure_1pc_copy.out:
```diff
@@ -0,0 +1,107 @@
+(1 row)
+
+COPY copy_test FROM PROGRAM 'echo 0, 0 && echo 1, 1 && echo 2, 4 && echo 3, 9' WITH CSV;
+ERROR:  failed to COPY to shard 100400 on localhost:57640
```
And again here, it the COPY fails while some rows are being sent we error out, instead of marking the placement inactive.

onderkalaci commented:
> An example might be, maybe if a packet comes in before the process is ready for it the process crashes. 

And if we really really want to do a test such a case, we could use isolation tester. 

> That class of bugs seems rare to me, so I think this approach is safe to move forward with.

I agree with that. I think those are not the type of tests we do in the release testing anyway.

onderkalaci commented on src/test/regress/expected/failure_drop_table.out:
```diff
@@ -0,0 +1,234 @@
+-- prove no network traffic happend
+SELECT count(1) FROM citus.mitmproxy('recorder.dump()');
+ count 
+-------
```
why do we see `1` if we're to prove no network traffic happened?

lithp replied:
This is a bug I haven't gotten around to fixing yet, it always returns an empty line at the end.

lithp replied:
Fixed: https://github.com/citusdata/citus/pull/2119/commits/a7e26cf44001ef1991dc0b59a9beccd611d5361f

onderkalaci commented:
> Can you add any test cases for this: #2031

It should be as simple as:
```
SELECT ... id = 100;
kill.flow()
SELECT ... id = 100;
```

lithp commented on src/test/regress/expected/failure_1pc_copy_hash.out:
```diff
@@ -0,0 +1,374 @@
+     4
+(1 row)
+
+SELECT citus.mitmproxy('flow.matches(b"COMMIT[^T]").killall()'); -- don't match "COMMITTED"
```
Here's a sample trace which shows what I mean:

````
SELECT citus.mitmproxy('recorder.reset()');
 mitmproxy 
-----------
 
(1 row)

COPY copy_test FROM PROGRAM 'echo 0, 0 && echo 1, 1 && echo 2, 4 && echo 3, 9' WITH CSV;
SELECT citus.mitmproxy('recorder.dump()');
                                                                                                                                                                                                                                                              
                                              mitmproxy                                                                                                                                                                                                       
                                                                                                      
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------------
 
 message from client: \x00\x00\x00W\x00\x03\x00\x00user\x00postgres\x00database\x00regression\x00application_name\x00citus\x00client_encoding\x00UTF8\x00\x00
 message from server: R\x00\x00\x00\x08\x00\x00\x00\x00S\x00\x00\x00\x1bapplication_name\x00citus\x00S\x00\x00\x00\x19client_encoding\x00UTF8\x00S\x00\x00\x00\x17DateStyle\x00ISO, MDY\x00S\x00\x00\x00\x19integer_datetimes\x00on\x00S\x00\x00\x00\x1bInterv
alStyle\x00postgres\x00S\x00\x00\x00\x14is_superuser\x00on\x00S\x00\x00\x00\x19server_encoding\x00UTF8\x00S\x00\x00\x00\x18server_version\x0010.3\x00S\x00\x00\x00#session_authorization\x00postgres\x00S\x00\x00\x00#standard_conforming_strings\x00on\x00S\x
00\x00\x00\x18TimeZone\x00US/Pacific\x00K\x00\x00\x00\x0c\x00\x00\x19\xbd7r\x0e\xa6Z\x00\x00\x00\x05I
 message from client: Q\x00\x00\x00qSELECT worker_apply_shard_ddl_command (100400, 'CREATE TABLE public.copy_test (key integer, value integer)')\x00
 message from server: T\x00\x00\x007\x00\x01worker_apply_shard_ddl_command\x00\x00\x00\x00\x00\x00\x00\x00\x00\x08\xe6\x00\x04\xff\xff\xff\xff\x00\x00D\x00\x00\x00\n\x00\x01\x00\x00\x00\x00C\x00\x00\x00\rSELECT 1\x00Z\x00\x00\x00\x05I
 message from client: Q\x00\x00\x00\x86BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 68, '2018-05-24 17:05:10.08712-07');\x00
 message from server: C\x00\x00\x00\nBEGIN\x00T\x00\x00\x00:\x00\x01assign_distributed_transaction_id\x00\x00\x00\x00\x00\x00\x00\x00\x00\x08\xe6\x00\x04\xff\xff\xff\xff\x00\x00D\x00\x00\x00\n\x00\x01\x00\x00\x00\x00C\x00\x00\x00\rSELECT 1\x00Z\x00\x00\x
00\x05T
 message from client: Q\x00\x00\x00ACOPY public.copy_test_100400 FROM STDIN WITH (FORMAT BINARY)\x00
 message from server: G\x00\x00\x00\x0b\x01\x00\x02\x00\x01\x00\x01
 message from client: d\x00\x00\x00\x17PGCOPY\n\xff\r\n\x00\x00\x00\x00\x00\x00\x00\x00\x00d\x00\x00\x00\x16\x00\x02\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00d\x00\x00\x00\x16\x00\x02\x00\x00\x00\x04\x00\x00\x00\x01\x00\x00\x00\x04\
x00\x00\x00\x01d\x00\x00\x00\x16\x00\x02\x00\x00\x00\x04\x00\x00\x00\x02\x00\x00\x00\x04\x00\x00\x00\x04d\x00\x00\x00\x16\x00\x02\x00\x00\x00\x04\x00\x00\x00\x03\x00\x00\x00\x04\x00\x00\x00\td\x00\x00\x00\x06\xff\xffc\x00\x00\x00\x04
 message from server: C\x00\x00\x00\x0bCOPY 4\x00Z\x00\x00\x00\x05T
 message from client: Q\x00\x00\x004SELECT pg_table_size('public.copy_test_100400')\x00
 message from server: T\x00\x00\x00&\x00\x01pg_table_size\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x14\x00\x08\xff\xff\xff\xff\x00\x00D\x00\x00\x00\x0e\x00\x01\x00\x00\x00\x048192C\x00\x00\x00\rSELECT 1\x00Z\x00\x00\x00\x05T
 message from client: Q\x00\x00\x00;SELECT min(key), max(key) FROM public.copy_test_100400\x00
 message from server: T\x00\x00\x002\x00\x02min\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x17\x00\x04\xff\xff\xff\xff\x00\x00max\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x17\x00\x04\xff\xff\xff\xff\x00\x00D\x00\x00\x00\x10\x00\x02\x00\x00\x00\x010\x00\x00\x
00\x013C\x00\x00\x00\rSELECT 1\x00Z\x00\x00\x00\x05T
 message from client: Q\x00\x00\x00\x0bCOMMIT\x00
 message from server: C\x00\x00\x00\x0bCOMMIT\x00Z\x00\x00\x00\x05I
 message from client: X\x00\x00\x00\x04
(18 rows)
````

If you just match on `COMMIT` then you'll preemptively kill the connection when `BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED` is sent, this regex excludes `COMMITTED` so we let that first packet though.

This is something I've been wondering about, I think these test are difficult to read unless you already know what the packet flow is, which, nobody does. It'd be great if we could include a packet dump at the top of the test which shows all the different places we're about to fail the command, but these dumps aren't reproducible, they change slightly every time.

Do you agree that would be valuable to have? I've been thinking of looking into sanitizing these dumps so we can embed them in test output.

lithp commented:
@onderkalaci cancellation [is now possible](https://github.com/citusdata/citus/pull/2119/commits/53aa3242fb4308610c40e78ee743b51f9d7a79e9#diff-bdc72553e1b3f3744316e4e001dc205bR133) :)

https://github.com/citusdata/citus/blob/53aa3242fb4308610c40e78ee743b51f9d7a79e9/src/test/regress/expected/failure_1pc_copy_append.out#L133-L154

lithp commented:
> I can see that we'll end up with 100s of tests soon. Could we parallelize the tests? Say, we start four proxies and divide the tests into four buckets via four pipes?

Interesting idea! This is hard because the proxy's port has to be added to pg_dist_node. We could abuse `citus.current_cluster`, and have a cluster per proxy, where the port for that cluster points at the correct proxy.

Dividing the tests into four buckets would be hard to maintain. In every test you would have to specify which pipe you want to talk to, and we'd have to be careful not not parallelize two tests which used the same pipe or else the tests would lock up.

Maybe we could spin up a new proxy for each test? We could add a `citus.startproxy()` function and a citus.releaseproxy()` function which finds a free port and spins one up and configures `citus.mitmfifo` and `citus.current_cluster`.

Or maybe we could do something fancy... say there are 4 proxies you're already spin up, you could add a `citus.claimproxy()` and a `citus.releaseproxy()` the first one blocks until a proxy is available and then configures everything for you. Just for testing we add a table, pg_dist_proxies, which has all the relevant information.

lithp commented on src/test/regress/mitmscripts/kill_after_begin.py:
```diff
@@ -0,0 +1,49 @@
+
+    print_message(tcp_msg)
+
+    # 2. Look for BEGIN. If you find it... mark the flow to be killed
```
Yeah, I need to delete this file

lithp commented on src/test/regress/input/failure_real_time_executor.source:
```diff
@@ -0,0 +1,99 @@
+SELECT citus.mitmproxy('flow.allow()');
```
Yeah, I agree that .source files are the worst. I wanted to duplicate the tests we already use in other places, I didn't want to have to figure out how to write my own join test, but you're right that I should just do it.

lithp commented on src/test/regress/expected/failure_1pc_copy_hash.out:
```diff
@@ -0,0 +1,374 @@
+CONTEXT:  COPY copy_test, line 1: "0, 0"
+-- ==== kill the connection when we try to start the COPY ====
+-- the query should abort
+SELECT citus.mitmproxy('flow.contains(b"FROM STDIN WITH").killall()');
```
> Do we always need to execute SELECT citus.mitmproxy('flow.allow()'); before any new killall or is it safe as we do here? '

We need a `flow.allow()` after any call to `flow.[things].killall()`, because once `killall` triggers it prevents any more connections from being made. If the last call was `kill` then we're safe, because only connections which include the thing `kill` matches against will be killed.

> Another question: Do we have anything like flow.disallow() or such to test the case where the coordinator cannot reach to the worker at all? Is that killall ?

Yep!

That's `flow.kill()` or `flow.killall()`, both do the same thing.

lithp replied:
`flow.reset()` probably does a better job of simulating a worker which can't be reached at all, but I don't think it makes a difference for us, `libpq` makes all the failure modes look the same.

lithp commented on src/test/regress/expected/failure_1pc_copy_hash.out:
```diff
@@ -0,0 +1,374 @@
+ 
+(1 row)
+
+COPY copy_test FROM PROGRAM 'echo 0, 0 && echo 1, 1 && echo 2, 4 && echo 3, 9' WITH CSV;
```
What do you mean? I think this format is the new windows-approved COPY format.

lithp commented on src/backend/distributed/citus--7.4-3--7.4-4.sql:
```diff
@@ -0,0 +1,18 @@
+
+  INSERT INTO mitmproxy_command VALUES (command);
+
+  EXECUTE format('COPY mitmproxy_command TO %L', current_setting('citus.mitmfifo'));
```
Yeah, `citus.mitmfifo` is set by either `pg_regress_multi.pl` or the user (when testing manually). It's "mitmproxy.fifo" but you need to give COPY an absolute path.

lithp commented on src/backend/distributed/Makefile:
```diff
@@ -199,6 +199,8 @@ $(EXTENSION)--7.4-2.sql: $(EXTENSION)--7.4-1.sql $(EXTENSION)--7.4-1--7.4-2.sql
 	cat $^ > $@
 $(EXTENSION)--7.4-3.sql: $(EXTENSION)--7.4-2.sql $(EXTENSION)--7.4-2--7.4-3.sql
 	cat $^ > $@
+$(EXTENSION)--7.4-4.sql: $(EXTENSION)--7.4-3.sql $(EXTENSION)--7.4-3--7.4-4.sql
```
Good point, although I've realized that this shouldn't be a version bump at all, we never want someone to use `citus.mitmproxy`. I think we have a `.sql` file somewhere with test functions, I'll add it to that.

onderkalaci commented:
> @onderkalaci cancellation is now possible :)

This is great! Hope `time.sleep(0.1)` wouldn't be a problem in windows 😄

onderkalaci commented:
> Just for testing we add a table, pg_dist_proxies, which has all the relevant information.

Hmm, I think to make things simpler in the first iteration, let's not try to parallelize things?

Actually, what I've in mind is that, let's pick a few tests, which covers many different types of commands, from the failure release test, and try to merge the PR with those tests. Later, ask anyone in the team to pick and implement 1-2 tests using the framework so that everyone becomes familiar with it. 

You or I could review those patches. What do you think about such an approach?

onderkalaci commented on src/test/regress/sql/failure_test_helpers.sql:
```diff
@@ -0,0 +1,47 @@
+CREATE FUNCTION citus.dump_network_traffic(
+  normalize_shards bool default true,
+  dump_unknown_messages bool default false
+) RETURNS TABLE(conn int, from_client bool, message text) AS $$
```
`from_client bool` use a text instead

lithp replied:
It's now `source text`, and a lot easier to read!

onderkalaci commented on src/test/regress/Makefile:
```diff
@@ -79,6 +79,10 @@ check-follower-cluster: all
 	$(pg_regress_multi_check) --load-extension=citus --follower-cluster \
 	-- $(MULTI_REGRESS_OPTS) --schedule=$(citus_abs_srcdir)/multi_follower_schedule $(EXTRA_TESTS)
 
+check-failure: all
```
If simple: stderr is directed to a file instead of stdout

onderkalaci replied:
already done

lithp commented:
I've rewritten history so that this branch never messed with the citus version or added a .sql file for it, and instead always used `test_helpers.sql`. This made rebasing easy, and I've now rebased it onto master. @onderkalaci you should now be able to test sequential modifications!

lithp commented:
I've removed the changes to prepared transaction id generation and reworked the test, `failure_drop_table`, to no longer rely on it. I've pushed the changes to [a branch](https://github.com/citusdata/citus/tree/unique-prepared-transaction-ids) for posterity in case we want them back, I think the test is a lot easier to read with them.

onderkalaci commented:
>  @onderkalaci you should now be able to test sequential modifications!

All looks good, I've added the tests in #2212. Once you address the feedback we've added together [here](https://github.com/citusdata/citus/pull/2119#pullrequestreview-127910395), feel free to ping me. I'd like to have a final look to the changes.

> Add tests for DDLs and real-time SELECTs [Onder]

I think I won't have time to add proper regression tests for `real-time` select failure / cancellation right now. I'm planning to tackle those during the release testing. For the other tests that are in the `failure_schedule`, feel free to pick one and add tests covering various edge cases. I'm OK to remove the other tests and let the whole team involve in writing all the tests mentioned here. 

We should probably get #2212 and #2210 to master (plus one other extensive test written by you).

lithp commented:
I've merged #2182 and rebased onto master so that this PR now includes 0 changes to citus code

lithp commented on src/test/regress/mitmscripts/fluent.py:
```diff
@@ -0,0 +1,477 @@
+import structs
+
+'''
+Use with a command line like this:
```
done

lithp commented on src/test/regress/mitmscripts/fluent.py:
```diff
@@ -0,0 +1,477 @@
+There are also some special commands. This proxy also records every packet and lets you
+inspect them:
+
+recorder.dump() - emits a list of captured packets in COPY text format
```
also done

lithp commented on src/test/regress/pg_regress_multi.pl:
```diff
@@ -275,6 +287,18 @@ sub revert_replace_postgres
 push(@pgOptions, '-c', "citus.remote_task_check_interval=1ms");
 push(@pgOptions, '-c', "citus.shard_replication_factor=2");
 push(@pgOptions, '-c', "citus.node_connection_timeout=${connectionTimeout}");
+push(@pgOptions, '-c', "citus.sslmode=disable");
```
done

lithp commented on src/test/regress/pg_regress_multi.pl:
```diff
@@ -178,6 +185,11 @@ ()
 MESSAGE
 }
 
+if ($useMitmproxy)
```
I've added `check-failure` to `check-full`.

lithp replied:
I've also added a line item to the travis build matrix, on PG10 we test both the regular tests and also the failure tests

lithp commented on src/test/regress/mitmscripts/fluent.py:
```diff
@@ -0,0 +1,477 @@
+        with open(fifoname, mode='w') as fifo:
+            fifo.write('{}\n'.format(result))
+
+def replace_thread(fifoname):
```
done

lithp commented on src/test/regress/mitmscripts/fluent.py:
```diff
@@ -0,0 +1,477 @@
+        self.root = self
+        self.command = None
+
+    def dump(self, normalize_shards=True, dump_unknown_messages=False):
```
removed

lithp commented:
@onderkalaci ready for you to review again :)

I think this is my only remaining work:
- send mitmproxy's stderr to a file instead of the console
- rewrite the tests to use the newer API (I think there are enough tests left for the team to write that it's not a problem to leave some of them in here)
- review and merge #2212 and #2210

lithp commented on src/test/regress/mitmscripts/README:
```diff
@@ -0,0 +1,165 @@
+Automated Failure testing
```
:grinning:

lithp commented on src/test/regress/mitmscripts/fluent.py:
```diff
@@ -0,0 +1,423 @@
+    def _handle(self, flow, message):
+        flow.kill() # tell mitmproxy this connection should be closed
+
+        client_conn = flow.client_conn # connections.ClientConnection(tcp.BaseHandler)
```
I can remove this, I think I added this while I was figuring out how to get the actual socket. This is just a note so I could remember what the type is `client_conn` is.

lithp replied:
Changed the comment to something more clear

lithp commented on src/test/regress/Pipfile.lock:
```diff
@@ -0,0 +1,328 @@
+{
```
I don't fully understand the reasoning but it sounds like adding `Pipfile.lock` is recommended: https://github.com/pypa/pipenv/issues/598#issuecomment-372941459

jasonmp85 replied:
> I don't fully understand the reasoning

Having just read waaaay too much about the craptastic Python dependency-management universe (I do not understand why it is _so_ Balkanized), we should _definitely_ check this in. `Pipfile` has the general version constraints and `Pipfile.lock` has the specific versions that we're using which meet those constraints. We'd only want to not provide `Pipfile.lock` if we were developing a library, I think.

lithp commented:
@onderkalaci addressed your feedback, removed all the tests (I'll open them again in new PRs, already opened #2244), sent all output to a file, this is ready for you again!

onderkalaci commented on src/test/regress/mitmscripts/README:
```diff
@@ -0,0 +1,165 @@
+
+# II. Running mitmproxy manually
+
+$ mkfifo /tmp/mitm.fifo  # first, you need a fifo
```
Maybe add ` cd src/test/regress/mitmscripts` ?

onderkalaci commented on src/test/regress/mitmscripts/README:
```diff
@@ -0,0 +1,165 @@
+means mitmdump will accept connections on port 9702 and forward them to the worker
+listening on port 9700.
+
+Now, open psql and run:
```
maybe add `cd src/test/regress`

onderkalaci commented on src/test/regress/mitmscripts/README:
```diff
@@ -0,0 +1,165 @@
+You also want to tell the UDFs how to talk to mitmproxy (careful, this must be an absolute
+path):
+
+# SET citus.mitmfifo = '/tmp/mitm.fifo';
```
Shall we note that this is not actually a GUC, but, still Postgres allows us to read it? Though I'm now familiar with this, I was confused when I first saw/realize this fact.

lithp replied:
Sure, I can add a quick note

metdos commented:
@lithp, when are we expecting to merge this?

lithp commented on .travis.yml:
```diff
@@ -1,6 +1,8 @@
 sudo: required
 dist: trusty
 language: c
+python:
```
Hmm, interesting. This script hooks unfortunately deeply into the mitmproxy, which is why the version requirement is so specific.

One solution might be to move test-automation from 2.7 to 3.5.

Another would be to replace mitmproxy with our own proxy, we're not using very much of mitmproxy so this wouldn't be a huge change, probably only a week.

jasonmp85 replied:
I've been using `pyenv` to install other Python versions and using `pipenv` on top of that to isolate all this. This works reliably well now.

metdos commented:
> @metdos @lithp so we merged this but that checklist looks pretty un-checked. Are we going to track further outstanding work elsewhere?

I added test conversion parts to this issue (#2262), do we have any plans for others @lithp?
